core/services/pipeline/orm.go

```diff
@@ -51,6 +58,139 @@ func (o *orm) CreateSpec(ctx context.Context, tx *gorm.DB, pipeline Pipeline, ma
+				return err
+			}
+		} else {
+			// Simply finish the run, no need to do any sort of locking
+			if run.Outputs.Val == nil || len(run.Errors) == 0 {
+				return errors.Errorf("run must have both Outputs and Errors, got Outputs: %#v, Errors: %#v", run.Outputs.Val, run.Errors)
+			}
+			sql := `UPDATE pipeline_runs SET state = :state, finished_at = :finished_at, errors= :errors, outputs = :outputs WHERE id = :id`
+			if _, err := tx.NamedExec(sql, run); err != nil {
+				return err
+			}
+		}
+
+		sql := `
+		INSERT INTO pipeline_task_runs (pipeline_run_id, task_run_id, type, index, output, error, dot_id, created_at, finished_at)
+		VALUES (:pipeline_run_id, :task_run_id, :type, :index, :output, :error, :dot_id, :created_at, :finished_at)
+		ON CONFLICT (pipeline_run_id, dot_id) DO UPDATE SET
+		output = EXCLUDED.output, error = EXCLUDED.error, finished_at = EXCLUDED.finished_at
+		RETURNING *;
+		`
+
+		// NOTE: can't use Select() to auto scan because we're using NamedQuery,
+		// sqlx.Named + Select is possible but it's about the same amount of code
+		rows, err := tx.NamedQuery(sql, run.PipelineTaskRuns)
+		if err != nil {
+			return err
+		}
+		taskRuns := []TaskRun{}
+		if err := sqlx.StructScan(rows, &taskRuns); err != nil {
+			return err
+		}
+		// replace with new task run data
+		run.PipelineTaskRuns = taskRuns
+		return nil
+	})
+	return restart, err
+}
+
+func (o *orm) UpdateTaskRun(db *sql.DB, taskID uuid.UUID, result interface{}) (run Run, start bool, err error) {
+	err = postgres.SqlxTransaction(context.Background(), db, func(tx *sqlx.Tx) error {
+		sql := `
+		SELECT pipeline_runs.* FROM pipeline_runs, pipeline_task_runs
+		WHERE pipeline_task_runs.task_run_id = $1
+		FOR UPDATE `
+		if err = tx.Get(&run, sql, taskID); err != nil {
+			return err
+		}
+
+		// Update the task with result
+		sql = `UPDATE pipeline_task_runs SET output = $2, finished_at = $3 WHERE task_run_id = $1`
+		if _, err = tx.Exec(sql, taskID, JSONSerializable{Val: result}, time.Now()); err != nil {
+			return err
+		}
+
+		if run.State == RunStatusSuspended {
+			start = true
+			run.State = RunStatusRunning
+
+			// We're going to restart the run, so set it back to "in progress"
+			sql = `UPDATE pipeline_runs SET state = $2 WHERE id = $1`
+			if _, err = tx.Exec(sql, run.ID, run.State); err != nil {
+				return err
+			}
+
+			// NOTE: can't join and preload in a single query unless explicitly listing all the struct fields...
+			// https://snippets.aktagon.com/snippets/757-how-to-join-two-tables-with-jmoiron-sqlx
+			sql = `SELECT * FROM pipeline_task_runs WHERE pipeline_run_id = $1`
+			if err = tx.Select(&run.PipelineTaskRuns, sql, run.ID); err != nil {
+				return err
+			}
+
+			// TODO: what if the node is killed while a run is in the state of running?
+			// we need to have a init goroutine that will sweep the db for any jobs still stuck on "running" at boot
```

Comment:
seems like a blocker right?

Reply:
Added in https://github.com/smartcontractkit/chainlink/pull/4571/commits/b953d7dc9f30db3a4e6bef1f9d402793738a573d